File: PRACTICA_JAVA/src/irrgarten/TestP1.java
package irrgarten;

import java.util.Arrays;

public class TestP1 {

    public static void main(String[] args) {
        // 1. Imprimir valores enumerados
        System.out.println("=== Inicio de pruebas de la práctica ===");

        System.out.println("\nValores enumerados:");
        System.out.println("Irrgarten.GameCharacter.PLAYER    => " + GameCharacter.PLAYER);
        System.out.println("Irrgarten.GameCharacter.MONSTER   => " + GameCharacter.MONSTER);
        System.out.println("Irrgarten.Orientation.VERTICAL    => " + Orientation.VERTICAL);
        System.out.println("Irrgarten.Orientation.HORIZONTAL  => " + Orientation.HORIZONTAL);
        System.out.println("Irrgarten.Directions.LEFT         => " + Directions.LEFT);
        System.out.println("Irrgarten.Directions.RIGHT        => " + Directions.RIGHT);
        System.out.println("Irrgarten.Directions.UP           => " + Directions.UP);
        System.out.println("Irrgarten.Directions.DOWN         => " + Directions.DOWN);

        // 2. Crear instancias de las clases
        System.out.println("\nProbar las instancias de las clases.");

        // Prueba de GameState
        System.out.println("\nClase GameState:");
        GameState gameState = new GameState("Laberinto1", "Jugador1, Jugador2", "Monstruo1", 0, false, "");
        System.out.println("Labyrinth: " + gameState.getLabyrinth());
        System.out.println("Players: " + gameState.getPlayers());
        System.out.println("Monsters: " + gameState.getMonsters());
        System.out.println("Current Player: " + gameState.getCurrentPlayer());
        System.out.println("Winner: " + gameState.getWinner());
        System.out.println("Log: " + gameState.getLog());

        // Prueba de Shield
        System.out.println("\nPrueba de Shield:");
        Shield shield = new Shield(5.0f, 3);
        System.out.println("Antes de usar protect: " + shield);
        float protection = shield.protect();
        System.out.println("Resultado de protect: " + protection);
        System.out.println("Después de usar protect: " + shield);
        boolean discardShield = shield.discard();
        System.out.println("Resultado de discard: " + discardShield);

        // Prueba de Weapon
        System.out.println("\nPrueba de Weapon:");
        Weapon weapon = new Weapon(7.0f, 4);
        System.out.println("Antes de usar attack: " + weapon);
        float damage = weapon.attack();
        System.out.println("Resultado de attack: " + damage);
        System.out.println("Después de usar attack: " + weapon);
        boolean discardWeapon = weapon.discard();
        System.out.println("Resultado de discard: " + discardWeapon);

        // 3. Prueba de métodos de Dice (100 iteraciones cada uno)
        Dice dice = new Dice();
        System.out.println("\nPrueba de métodos de Dice (100 iteraciones cada uno):");

        // randomPos(10)
        System.out.println("\nDistribución de random_pos(10):");
        int[] randomPosCount = new int[10];
        for (int i = 0; i < 100; i++) {
            randomPosCount[dice.randomPos(10)]++;
        }
        for (int i = 0; i < randomPosCount.length; i++) {
            System.out.println("Valor " + i + ": " + randomPosCount[i] + " veces");
        }

        // whoStarts(4)
        System.out.println("\nDistribución de who_starts(4):");
        int[] whoStartsCount = new int[4];
        for (int i = 0; i < 100; i++) {
            whoStartsCount[dice.whoStarts(4)]++;
        }
        for (int i = 0; i < whoStartsCount.length; i++) {
            System.out.println("Jugador " + i + ": " + whoStartsCount[i] + " veces");
        }

        // randomIntelligence()
        System.out.println("\nValor medio de random_intelligence: " + average(dice, 100, "randomIntelligence"));

        // randomStrength()
        System.out.println("Valor medio de random_strength: " + average(dice, 100, "randomStrength"));

        // resurrectPlayer
        System.out.println("resurrect_player ha devuelto true " + countTrue(dice, 100) + " de 100 veces");

        // weaponsReward()
        System.out.println("\nDistribución de weapons_reward:");
        int[] weaponsRewardCount = new int[3];
        for (int i = 0; i < 100; i++) {
            weaponsRewardCount[dice.weaponsReward()]++;
        }
        for (int i = 0; i < weaponsRewardCount.length; i++) {
            System.out.println("Reward " + i + ": " + weaponsRewardCount[i] + " veces");
        }

        // shieldsReward()
        System.out.println("\nDistribución de shields_reward:");
        int[] shieldsRewardCount = new int[4];
        for (int i = 0; i < 100; i++) {
            shieldsRewardCount[dice.shieldsReward()]++;
        }
        for (int i = 0; i < shieldsRewardCount.length; i++) {
            System.out.println("Reward " + i + ": " + shieldsRewardCount[i] + " veces");
        }

        // healthReward()
        System.out.println("\nDistribución de health_reward:");
        int[] healthRewardCount = new int[6];
        for (int i = 0; i < 100; i++) {
            healthRewardCount[dice.healthReward()]++;
        }
        for (int i = 0; i < healthRewardCount.length; i++) {
            System.out.println("Reward " + i + ": " + healthRewardCount[i] + " veces");
        }

        // weaponPower()
        System.out.println("\nDistribución de weapon_power:");
        int[] weaponPowerCount = new int[3];
        for (int i = 0; i < 100; i++) {
            weaponPowerCount[(int) dice.weaponPower()]++;
        }
        for (int i = 0; i < weaponPowerCount.length; i++) {
            System.out.println("Power " + i + ": " + weaponPowerCount[i] + " veces");
        }

        // shieldPower()
        System.out.println("\nDistribución de shield_power:");
        int[] shieldPowerCount = new int[2];
        for (int i = 0; i < 100; i++) {
            shieldPowerCount[(int) dice.shieldPower()]++;
        }
        for (int i = 0; i < shieldPowerCount.length; i++) {
            System.out.println("Power " + i + ": " + shieldPowerCount[i] + " veces");
        }

        // usesLeft()
        System.out.println("\nDistribución de uses_left:");
        int[] usesLeftCount = new int[6];
        for (int i = 0; i < 100; i++) {
            usesLeftCount[dice.usesLeft()]++;
        }
        for (int i = 0; i < usesLeftCount.length; i++) {
            System.out.println("Uses " + i + ": " + usesLeftCount[i] + " veces");
        }

        // intensity()
        System.out.println("\nValor medio de intensity(5): " + averageIntensity(dice, 100, 5));

        // discardElement(3)
        System.out.println("\nResultados de discard_element(3):");
        int trueCount = 0, falseCount = 0;
        for (int i = 0; i < 100; i++) {
            if (dice.discardElement(3)) {
                trueCount++;
            } else {
                falseCount++;
            }
        }
        System.out.println("true: " + trueCount + " veces");
        System.out.println("false: " + falseCount + " veces");

        System.out.println("\n=== Pruebas completadas ===");
    }

    // Método corregido para calcular el valor medio
    private static double average(Dice dice, int iterations, String methodName) {
        double sum = 0;
        for (int i = 0; i < iterations; i++) {
            try {
                Float value = (Float) Dice.class.getMethod(methodName).invoke(dice);
                sum += value.doubleValue();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return sum / iterations;
    }

    private static double averageIntensity(Dice dice, int iterations, float competence) {
        double sum = 0;
        for (int i = 0; i < iterations; i++) {
            sum += dice.intensity(competence);
        }
        return sum / iterations;
    }

    private static int countTrue(Dice dice, int iterations) {
        int count = 0;
        for (int i = 0; i < iterations; i++) {
            if (dice.resurrectPlayer()) {
                count++;
            }
        }
        return count;
    }
}

File: PRACTICA_RUBY/test_p1.rb
# encoding:utf-8
require_relative 'Dice'
require_relative 'GameState'
require_relative 'Shield'
require_relative 'Weapon'
require_relative 'GameCharacter'
require_relative 'Orientation'
require_relative 'Directions'


# Se asume que las clases y módulos (Dice, GameState, Shield, Weapon, GameCharacter, Orientation, Directions) ya han sido definidos previamente en la práctica.

class TestP1
  def self.main
    puts "=== Inicio de pruebas de la práctica ==="

    # --- Prueba de los tipos enumerados ---
    puts "\nValores enumerados:"
    puts "Irrgarten::GameCharacter::PLAYER    => #{Irrgarten::GameCharacter::PLAYER}"
    puts "Irrgarten::GameCharacter::MONSTER   => #{Irrgarten::GameCharacter::MONSTER}"
    puts "Irrgarten::Orientation::VERTICAL    => #{Irrgarten::Orientation::VERTICAL}"
    puts "Irrgarten::Orientation::HORIZONTAL  => #{Irrgarten::Orientation::HORIZONTAL}"
    puts "Irrgarten::Directions::LEFT         => #{Irrgarten::Directions::LEFT}"
    puts "Irrgarten::Directions::RIGHT        => #{Irrgarten::Directions::RIGHT}"
    puts "Irrgarten::Directions::UP           => #{Irrgarten::Directions::UP}"
    puts "Irrgarten::Directions::DOWN         => #{Irrgarten::Directions::DOWN}"

    # --- Creación de instancias de las clases de la práctica ---
    puts "\nCreando instancias de clases..."
    dice       = Irrgarten::Dice.new
    game_state = Irrgarten::GameState.new("Laberinto1", ["Jugador1", "Jugador2"], ["Monstruo1"], "Jugador1", nil, [])
    shield     = Irrgarten::Shield.new(5, 3)
    weapon     = Irrgarten::Weapon.new(7, 4)

    # --- Prueba de la clase GameState ---
    puts "\nPrueba de GameState:"
    puts "Labyrinth: #{game_state.labyrinth}"
    puts "Players: #{game_state.players}"
    puts "Monsters: #{game_state.monsters}"
    puts "Current Player: #{game_state.current_player}"
    puts "Winner: #{game_state.winner}"
    puts "Log: #{game_state.log}"

    # --- Prueba de la clase Shield ---
    puts "\nPrueba de Shield:"
    puts "Antes de usar protect: #{shield.to_s}"
    defensa = shield.protect
    puts "Resultado de protect: #{defensa}"
    puts "Después de usar protect: #{shield.to_s}"
    puts "Resultado de discard: #{shield.discard}"

    # --- Prueba de la clase Weapon ---
    puts "\nPrueba de Weapon:"
    puts "Antes de usar attack: #{weapon.to_s}"
    daño = weapon.attack
    puts "Resultado de attack: #{daño}"
    puts "Después de usar attack: #{weapon.to_s}"
    puts "Resultado de discard: #{weapon.discard}"

    # --- Prueba de la clase Dice 100 veces ---
    puts "\nPrueba de métodos de Dice (100 iteraciones cada uno):"
    test_dice_methods(dice)

    puts "\n=== Pruebas completadas ==="
  end

  # Método auxiliar para probar los métodos de Dice 100 veces y mostrar estadísticas básicas.
  def self.test_dice_methods(dice)
    num_trials = 100

    # random_pos(max)
    pos_distribution = Hash.new(0)
    num_trials.times { pos_distribution[dice.random_pos(10)] += 1 }
    puts "\nDistribución de random_pos(10):"
    pos_distribution.sort.each { |valor, veces| puts "Valor #{valor}: #{veces} veces" }

    # who_starts(nplayers)
    start_distribution = Hash.new(0)
    num_trials.times { start_distribution[dice.who_starts(4)] += 1 }
    puts "\nDistribución de who_starts(4):"
    start_distribution.sort.each { |jugador, veces| puts "Jugador #{jugador}: #{veces} veces" }

    # random_intelligence
    intel_values = Array.new(num_trials) { dice.random_intelligence }
    avg_intel = intel_values.sum / num_trials.to_f
    puts "\nValor medio de random_intelligence: #{avg_intel}"

    # random_strength
    strength_values = Array.new(num_trials) { dice.random_strength }
    avg_strength = strength_values.sum / num_trials.to_f
    puts "\nValor medio de random_strength: #{avg_strength}"

    # resurrect_player (cuenta cuántas veces devuelve true)
    resurrect_count = 0
    num_trials.times { resurrect_count += 1 if dice.resurrect_player }
    puts "\nresurrect_player ha devuelto true #{resurrect_count} de #{num_trials} veces"

    # weapons_reward
    weapons_rewards = Hash.new(0)
    num_trials.times { weapons_rewards[dice.weapons_reward] += 1 }
    puts "\nDistribución de weapons_reward:"
    weapons_rewards.sort.each { |reward, veces| puts "Reward #{reward}: #{veces} veces" }

    # shields_reward
    shields_rewards = Hash.new(0)
    num_trials.times { shields_rewards[dice.shields_reward] += 1 }
    puts "\nDistribución de shields_reward:"
    shields_rewards.sort.each { |reward, veces| puts "Reward #{reward}: #{veces} veces" }

    # health_reward
    health_rewards = Hash.new(0)
    num_trials.times { health_rewards[dice.health_reward] += 1 }
    puts "\nDistribución de health_reward:"
    health_rewards.sort.each { |reward, veces| puts "Reward #{reward}: #{veces} veces" }

    # weapon_power
    weapon_powers = Hash.new(0)
    num_trials.times { weapon_powers[dice.weapon_power] += 1 }
    puts "\nDistribución de weapon_power:"
    weapon_powers.sort.each { |power, veces| puts "Power #{power}: #{veces} veces" }

    # shield_power
    shield_powers = Hash.new(0)
    num_trials.times { shield_powers[dice.shield_power] += 1 }
    puts "\nDistribución de shield_power:"
    shield_powers.sort.each { |power, veces| puts "Power #{power}: #{veces} veces" }

    # uses_left
    uses_distribution = Hash.new(0)
    num_trials.times { uses_distribution[dice.uses_left] += 1 }
    puts "\nDistribución de uses_left:"
    uses_distribution.sort.each { |uses, veces| puts "Uses #{uses}: #{veces} veces" }

    # intensity(competence): Usamos competencia = 5
    intensity_values = Array.new(num_trials) { dice.intensity(5) }
    avg_intensity = intensity_values.sum / num_trials.to_f
    puts "\nValor medio de intensity(5): #{avg_intensity}"

    # discard_element(uses_left): Probamos con un valor fijo de uses_left, por ejemplo 3.
    discard_results = { true => 0, false => 0 }
    num_trials.times { discard_results[dice.discard_element(3)] += 1 }
    puts "\nResultados de discard_element(3):"
    discard_results.each { |resultado, veces| puts "#{resultado}: #{veces} veces" }
  end
end

# Al final del fichero se invoca el método main para ejecutar el programa principal.
TestP1.main
